pygfx/objects/_ruler.py

```diff
@@ -0,0 +1,489 @@
+                if maybe_step > approx_step:
+                    step = maybe_step
+                    break
+
+        return step
+
+    def _get_ticks_uniform(self, min_value, max_value, step):
+        """Get a uniformly distributed set of ticks."""
+
+        if not step:
+            return {}
+
+        tick_format = self._tick_format
+
+        # Apply some form of scaling
+        ref_value = max(abs(min_value), abs(max_value))
+        if tick_format.lower() == "km":  # use mk units
+            tick_format = "0.4g"  # for the remaining number
+            if ref_value >= 10_000_000_000:
+                mult, unit = 1 / 1_000_000_000, "G"
+            elif ref_value >= 10_000_000:
+                mult, unit = 1 / 1_000_000, "M"
+            elif ref_value >= 10000:
+                mult, unit = 1 / 1000, "K"
+            elif ref_value < 0.0001:
+                mult, unit = 1_000_000, "u"
+            elif ref_value < 0.1:
+                mult, unit = 1000, "m"
+            else:
+                mult, unit = 1, ""
+        else:
+            mult, unit = 1, ""
+
+        # Calculate tick values
+        first_tick = ceil(min_value / step) * step
+        last_tick = floor(max_value / step) * step
+        ticks = {}
+        t = first_tick
+        while t <= last_tick:
+            if t == 0:
+                s = "0"
+            else:
+                s = format(mult * t, tick_format)
+            ticks[t] = s + unit
+            t += step
+
+        return ticks
+
+    def _update_sub_objects(self, ticks, tick_auto_step):
+        """Update the sub-objects to show the given ticks."""
+        assert isinstance(ticks, dict)
+
+        tick_size = 5
+        min_n_slots = 8  # todo: can be (much) higher when we use a single text object!
+
+        # Load config
+        start_pos = self._start_pos
+        end_pos = self._end_pos
+        start_value = self._start_value
+        end_value = self.end_value
+
+        # Derive some more variables
+        length = end_value - start_value
+        vec = end_pos - start_pos
+        if length:
+            vec /= length
+
+        # Get array to store positions
+        n_slots = self.points.geometry.positions.nitems
+        n_positions = len(ticks) + 2
+        if n_positions <= n_slots <= max(min_n_slots, 2 * n_positions):
+            # Re-use existing buffers
+            positions = self.points.geometry.positions.data
+            sizes = self.points.geometry.sizes.data
+            self.points.geometry.positions.update_range()
+            self.points.geometry.sizes.update_range()
+        else:
+            # Allocate new buffers
+            new_n_slots = max(min_n_slots, int(n_positions * 1.2))
+            positions = np.zeros((new_n_slots, 3), np.float32)
+            sizes = np.zeros((new_n_slots,), np.float32)
+            self.points.geometry.positions = Buffer(positions)
+            self.points.geometry.sizes = Buffer(sizes)
+
+        def define_text(pos, text):
```

Comment:
does the pool ever shrink?

Reply:
Yes, if `n_positions` is smaller than `2 * n_slots`, and larger than `min_n_slots`.

Reply:
No wait, the pool is managed differently. There's ``self._text_object_pool[index + 1 :] = []`` some lines below. I will have a look at whether this can be combined better with the "pool" of lines and points.

Reply:
Done